### DISCUSSION TITLE
Render、EntityRender与PlayerAnimation相关的文档/教程/讨论/类询问。

### DISCUSSION THREAD

**Discussion by M1hono** [mouse0w0/MinecraftDeveloperGuide#90] [问答 / Q&A]
各位好，我最近在恶补标题相关的三个类别的知识以进行**实体相关**的开发。
<mark>首先对于Render，我希望了解是否Render需要了解的就是OpenGL相关，也就是本导航中有关Shader的部分，目前我对PoseStack非常困惑期望得到解惑，而直接查看原版类并一定地问答GPT后我仍旧有着一些担心影响到开发的疑惑，所以希望能够找到更加系统的教程。<mark>

> 当然，如果能够提供一些规范地实现了上述要素与使用了PoseStack的原版类也一定能够帮助到我，如果您恰巧对此有所了解的话希望能够获取到您的帮助。

Entity Render 和 PlayerRender是类似的问题，但涉及到一个关于Model的进一步的疑问，也就是有关Part的操作，由于我不想要借用"Player Animator"只想实现一些简单的人物手臂与身体的动作，如何操作Model Part并规范地作为EntityEvent使用就变成了我心中的疑惑。

> 如果可以的话Camera也想知晓具体的操作所在的类，虽然有着现成的库，但是我不想放弃诸如此类的学习机会所以。

**Comment by mouse0w0** [comment]
你的提问表述有些混乱，所以我有些难以知道你的问题是什么，仅根据我个人的理解来回答。

关于`PoseStack`，你需要了解图形学中[变换](https://learnopengl-cn.github.io/01%20Getting%20started/07%20Transformations/)和[坐标系统](https://learnopengl-cn.github.io/01%20Getting%20started/08%20Coordinate%20Systems/)的概念，并从中学习到以下四个概念：
- **变换**是**位移、旋转、缩放**这三个操作的统称。
- **局部坐标空间**指**模型自身定义的坐标空间**。
- **世界坐标空间**指物体在**游戏世界**上所处的坐标空间。
- `PoseStack`是一个用于存储**将物体的坐标从局部坐标空间变换到世界坐标空间**的矩阵的栈容器。

关于`ModelPart`，它字面意思即是指**一个模型的部分/部件**，而Minecraft的模型动画就是通过操作模型部件的**变换**来实现动画效果。

关于`Camera`，由于你没有给出对应的游戏版本和ModLoader信息，我无法回答，也不太清楚你究竟想要做什么。

个人建议先在[LearnOpenGL](https://learnopengl-cn.github.io/)和[lwjglbook](https://mouse0w0.github.io/lwjglbook-CN-Translation/)系统地学习OpenGL的相关基础知识，然后更加慎重的整理提问你的问题。

**Reply by M1hono** [reply]
确实是想要这种`具体应该学习什么`上的解惑，非常感谢，我会去研究这两个部分的相关实现。

Camera实际上是想要在Forge端实现一个类似基岩版的相机操作功能，来方便相关的开发。

是的，我的确不太熟悉这方面的内容……所以想要知道该如何开始。我会先去系统地了解了OpenGL的知识再慎重地考虑具体的需求，我先把这个讨论挂起，会在一个月内完成相关的学习。
